src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php

```diff
@@ -410,55 +410,78 @@ public function GetBasketSumForVoucher(TdbShopVoucher &$oVoucher)
     {
-        $iCurrentPosition = $this->getItemPointer();
+        $currentPosition = $this->getItemPointer();
         $this->GoToStart();
-        $oVoucherDef = &$oVoucher->GetFieldShopVoucherSeries();
-        $aArticlesAffected = array();
-        $dTotalValueOfAffectedItems = 0;
-        while ($oItem = &$this->Next()) {
-            $bIncludeProduct = $oVoucher->AllowVoucherForArticle($oItem);
-            if ($bIncludeProduct) {
-                $aArticlesAffected[] = &$oItem;
-                $dTotalValueOfAffectedItems = $dTotalValueOfAffectedItems + $oItem->dPriceTotalAfterDiscountWithoutVouchers;
+        $articlesAffected = array();
+        $totalValueOfAffectedItems = 0;
+        while ($basketItem = &$this->Next()) {
+            if (true === $oVoucher->AllowVoucherForArticle($basketItem)) {
+                $articlesAffected[] = &$basketItem;
+                $totalValueOfAffectedItems = $totalValueOfAffectedItems + $basketItem->dPriceTotalAfterDiscountWithoutVouchers;
             }
         }
-        $this->setItemPointer($iCurrentPosition);
+        $this->setItemPointer($currentPosition);
+
+        $this->correctBasketItemDiscountPrices($articlesAffected, $dVoucherValue, $totalValueOfAffectedItems);
+    }
 
-        $iNumberOfAffectedItems = count($aArticlesAffected);
-        $dTotalValueOfAffectedItemsUsed = 0;
+    /**
+     * @param TdbShopDiscount $discount
+     * @param float           $unusedDiscountValue value difference between discount based on single article and discount based on all articles
+     */
+    public function ReducePriceForItemsAffectedByDiscount(\TdbShopDiscount $discount, float $unusedDiscountValue): void
+    {
+        $currentPosition = $this->getItemPointer();
+        $this->GoToStart();
+        $articlesAffected = array();
+        $totalValueOfAffectedItems = 0;
+        while ($basketItem = &$this->Next()) {
+            if (true === $discount->AllowDiscountForArticle($basketItem)) {
+                $articlesAffected[] = &$basketItem;
+                $totalValueOfAffectedItems = $totalValueOfAffectedItems + $basketItem->dPriceTotalAfterDiscountWithoutVouchers;
+            }
+        }
+        $this->setItemPointer($currentPosition);
+
+        $this->correctBasketItemDiscountPrices($articlesAffected, $unusedDiscountValue, $totalValueOfAffectedItems);
+    }
+
+    protected function correctBasketItemDiscountPrices(array $articlesAffected, float $unusedDiscountValue, float $totalValueOfAffectedItems): void
+    {
+        $iNumberOfAffectedItems = count($articlesAffected);
         if ($iNumberOfAffectedItems > 0) {
-            $dRemainingVoucherUse = $dVoucherValue;
-            foreach (array_keys($aArticlesAffected) as $iAffectedArticleIndex) {
+            $dRemainingDiscountUse = $unusedDiscountValue;
+            foreach (array_keys($articlesAffected) as $iAffectedArticleIndex) {
```

Comment:
This method is really hard to read. Please improve it by
a) adding an early return if iNumberOfAffectedItems <= 0
b) replacing `foreach (array_keys($articlesAffected) as $iAffectedArticleIndex)` by `foreach ($articlesAffected as $iAffectedArticleIndex => $affectedArticle)` (and of course use $affectedArticle)
c) splitting the method in at least 3 methods

Thanks!

Reply:
imporve method realising a and b